### PR TITLE
Add option for healthz check port. Use it to disable server with installer option skip liveness probe.

### LINKED ISSUE #70: Be able to choose the healthcheck api-server port
Currently in Docker Desktop, we deploy Compose on Kubernetes with option `--network-host=true` so that we can reuse the same etcd as Kubernetes core API.

The default port for api-server is 8080 which collides with containers deployed by hand by users.

Steps:
1. Activate Kubernetes in Docker Desktop
2. `docker run -p 8080:80 -it nginx`

Output: `Error starting userland proxy: listen tcp 0.0.0.0:8080: bind: address already in use.`

When doing netstat inside the VM: `tcp        0      0 :::8080                 :::*                    LISTEN      8655/api-server` 

Is it possible to add a flag for this port in order to use a non-usual port and let 8080 free for our users ?

### PR DESCRIPTION
**Pull request by guillaumerose**
Fix #70

### CODE REVIEW

File: cmd/api-server/cli/root.go [hunk 1]
```diff
@@ -66,6 +67,7 @@ func NewCommandStartComposeServer(stopCh <-chan struct{}) *cobra.Command {
 	flags.StringVar(&o.serviceNamespace, "service-namespace", "", "defines the namespace of the service exposing the aggregated API")
 	flags.StringVar(&o.serviceName, "service-name", "", "defines the name of the service exposing the aggregated API")
 	flags.StringVar(&o.caBundleFile, "ca-bundle-file", "", "defines the path to the CA bundle file")
+	flags.IntVar(&o.healthzCheckPort, "healthz-check-port", 8080, "defines the port used by healthz check server")
```
Comment (simonferquel) [inline cmd/api-server/cli/root.go]:
```suggestion
	flags.IntVar(&o.healthzCheckPort, "healthz-check-port", 8080, "defines the port used by healthz check server (0 to disable it)")
```

File: install/install.go [hunk 2]
```diff
@@ -813,6 +816,10 @@ func (c *installer) createController(ctx *installerContext) error {
 		args = append(args, "--default-service-type="+c.commonOptions.DefaultServiceType)
 	}
 
+	if c.commonOptions.SkipLivenessProbes {
```
Comment (simonferquel) [inline install/install.go]: is it not already applied in `configureAPIServerArgs()` ?

Comment (ijc) [reply install/install.go]: I convinced myself those code paths didn't overlap - very possible I was wrong though!

Comment (guillaumerose) [reply install/install.go]: this one is only for the controller

Comment (simonferquel) [reply install/install.go]: oh, right

**Comment by guillaumerose** [comment]
I updated the PR according to comments. I tried to beat gocyclo by extracting the checkAPIPresent() method.

**Comment by silvin-lubecki** [comment]
@guillaumerose  the validate step is complaining:
```
cmd/compose-controller/main.go:1::warning: file is not goimported (goimports)
```

**Comment by silvin-lubecki** [comment]
e2e tests are failing too:
```
Compose fry with permission 
  Should allow user to create stack
  /go/src/github.com/docker/compose-on-kubernetes/e2e/rbac_test.go:32

STEP: Creating namespace "e2e-tests-compose-1261616337108845555" for this suite.
error occurred, waiting: stacks.compose.docker.com "app" is forbidden: User "employee" cannot create resource "stacks/composefile" in API group "compose.docker.com" in the namespace "e2e-tests-compose-1261616337108845555"
STEP: Destroying namespace "e2e-tests-compose-1261616337108845555" for this suite.
```

File: cmd/api-server/cli/root.go [hunk 3]
```diff
@@ -241,23 +243,29 @@ func runComposeServer(o *apiServerOptions, stopCh <-chan struct{}) error {
...
+
+	if o.healthzCheckPort > 0 {
+		err = server.GenericAPIServer.AddPostStartHook("start-compose-server-healthz-endpoint", func(context genericapiserver.PostStartHookContext) error {
+			m := http.NewServeMux()
+			healthz.InstallHandler(m, server.GenericAPIServer.HealthzChecks()...)
+			srv := &http.Server{
+				Addr:    fmt.Sprintf(":%d", o.healthzCheckPort),
+				Handler: m,
+			}
+			go func() {
```
Comment (silvin-lubecki) [inline cmd/api-server/cli/root.go]: nit: I think `go srv.ListenAndServe()` works fine too.

File: cmd/compose-controller/main.go [hunk 4]
```diff
@@ -6,6 +6,7 @@ import (
 	"flag"
 	"fmt"
 	"io"
+	"k8s.io/client-go/rest"
```
Comment (silvin-lubecki) [inline cmd/compose-controller/main.go]: This import is not at the right place

File: cmd/compose-controller/main.go [hunk 5]
```diff
@@ -157,12 +152,26 @@ func start(opts *controllerOptions) error {
...
+		go http.ListenAndServe(fmt.Sprintf(":%d", opts.healthzCheckPort), nil)
+	}
 	<-stop
 	return nil
 }
 
+func checkAPIPresent(config *rest.Config) error {
+	var err error
+	for i := 0; i < 60; i++ {
+		err = check.APIPresent(config)
```
Comment (silvin-lubecki) [inline cmd/compose-controller/main.go]: nit: `if err = check.APIPresent(config); err == nil{`

File: install/install.go [hunk 6]
```diff
@@ -617,28 +617,31 @@ func (c *installer) configureAPIServerImage() (image string, args []string, env
...
 
+func (c *installer) configureAPIServerArgs() []string {
+	args := []string{
+		"--kubeconfig", "",
+		"--authentication-kubeconfig=",
+		"--authorization-kubeconfig=",
+		"--service-name=compose-api",
+		"--service-namespace", c.commonOptions.Namespace,
+	}
+	if c.commonOptions.SkipLivenessProbes {
```
Comment (silvin-lubecki) [inline install/install.go]: Or you can also have something like this at top level:
```go
if c.commonOptions.SkipLivenessProbes {
    c.commonOptions.HealthzCheckPort = 0
}
```
So here you would only have
```go
return append(args, fmt.Sprintf("--healthz-check-port=%d", c.commonOptions.HealthzCheckPort))
```

Comment (guillaumerose) [reply install/install.go]: fixed

File: cmd/compose-controller/main.go (hunk 5 above)
Comment (guillaumerose) [reply cmd/compose-controller/main.go]: fixed

File: cmd/compose-controller/main.go (hunk 4 above)
Comment (guillaumerose) [reply cmd/compose-controller/main.go]: fixed

File: cmd/api-server/cli/root.go (hunk 3 above)
Comment (guillaumerose) [reply cmd/api-server/cli/root.go]: fixed

**Comment by simonferquel** [comment]
Hmm, this changes the API server/controller behavior, but if I am not wrong, I think the deployment specs still reference port 8080 explicitly

**Comment by guillaumerose** [comment]
Oh yes, thanks

**Comment by simonferquel** [comment]
See https://github.com/docker/compose-on-kubernetes/blob/3b5db76eaff9253e1727ded79896c7ef1b2081ae/install/install.go#L849 and https://github.com/docker/compose-on-kubernetes/blob/3b5db76eaff9253e1727ded79896c7ef1b2081ae/install/install.go#L682

**Comment by guillaumerose** [comment]
Yes I just changed it.

**Comment by simonferquel** [comment]
https://github.com/docker/compose-on-kubernetes/pull/72/files#diff-cdae8f79a5c22908bba7c1ba91c8a357R697 -> test based on skipLivenessProbe option should now check for LivenessProbe port instead

**Comment by guillaumerose** [comment]
I removed `SkipLivenessProbes` and just kept `HealthzCheckPort`. The flag is still here and set the port to 0 if used.

**Comment by simonferquel** [comment]
This is merged, thanks!

**Pull request merged**